engines/sherlock/graphics.cpp

```diff
@@ -0,0 +1,205 @@
+	setPixels(src.getBasePtr(r.left, r.top));
+	w = r.width();
+	h = r.height();
+	pitch = src.pitch;
+	format = Graphics::PixelFormat::createFormatCLUT8();
+}
+
+Surface::Surface() : _freePixels(false) {
+}
+
+Surface::~Surface() {
+	if (_freePixels)
+		free();
+}
+
+/**
+ * Sets up an internal surface with the specified dimensions that will be automatically freed
+ * when the surface object is destroyed
+ */
+void Surface::create(uint16 width, uint16 height) {
+	if (_freePixels)
+		free();
+
+	Graphics::Surface::create(width, height, Graphics::PixelFormat::createFormatCLUT8());
+	_freePixels = true;
+}
+
+/**
+ * Copy a surface into this one
+ */
+void Surface::blitFrom(const Graphics::Surface &src) {
+	blitFrom(src, Common::Point(0, 0));
+}
+
+/**
+ * Draws a surface at a given position within this surface
+ */
+void Surface::blitFrom(const Graphics::Surface &src, const Common::Point &pt) {
+	Common::Rect drawRect(0, 0, src.w, src.h);
+	Common::Point destPt = pt;
+
+	if (destPt.x < 0) {
+		drawRect.left += -destPt.x;
+		destPt.x = 0;
+	}
+	if (destPt.y < 0) {
+		drawRect.top += -destPt.y;
+		destPt.y = 0;
+	}
+	int right = destPt.x + src.w;
+	if (right > this->w) {
+		drawRect.right -= (right - this->w);
+	}
+	int bottom = destPt.y + src.h;
+	if (bottom > this->h) {
+		drawRect.bottom -= (bottom - this->h);
+	}
+
+	if (drawRect.isValidRect())
+		blitFrom(src, destPt, drawRect);
+}
+
+/**
+ * Draws a sub-section of a surface at a given position within this surface
+ */
+void Surface::blitFrom(const Graphics::Surface &src, const Common::Point &pt,
+		const Common::Rect &srcBounds) {
+	Common::Rect destRect(pt.x, pt.y, pt.x + srcBounds.width(),
+		pt.y + srcBounds.height());
+	Common::Rect srcRect = srcBounds;
+
+	if (clip(srcRect, destRect)) {
+		// Surface is at least partially or completely on-screen
+		addDirtyRect(destRect);
+		copyRectToSurface(src, destRect.left, destRect.top, srcRect);
+	}
+}
+
+/**
+* Draws an image frame at a given position within this surface with transparency
+*/
+void Surface::transBlitFrom(const ImageFrame &src, const Common::Point &pt,
+		bool flipped, int overrideColor) {
+	transBlitFrom(src._frame, pt + src._offset, flipped, overrideColor);
+}
+
+/**
+* Draws a surface at a given position within this surface with transparency
+*/
+void Surface::transBlitFrom(const Graphics::Surface &src, const Common::Point &pt,
+		bool flipped, int overrideColor) {
+	Common::Rect drawRect(0, 0, src.w, src.h);
+	Common::Rect destRect(pt.x, pt.y, pt.x + src.w, pt.y + src.h);
+
+	// Clip the display area to on-screen
+	if (!clip(drawRect, destRect))
+		// It's completely off-screen
+		return;
+
+	if (flipped)
+		drawRect = Common::Rect(src.w - drawRect.right, src.h - drawRect.bottom,
+			src.w - drawRect.left, src.h - drawRect.top);
+
+	Common::Point destPt(destRect.left, destRect.top);
+	addDirtyRect(Common::Rect(destPt.x, destPt.y, destPt.x + drawRect.width(),
+		destPt.y + drawRect.height()));
```

Comment:
If the two rectangles returned by clip() have the same width, I think this should in fact always be equal to destRect.